Feature/week7 homework

Bartozo opened the pull request:
My finished homework.

It uses retrofit to download facts about the cats and store them in the room database.
Users can scroll and read every funny fact! (Api has no endpoint for voting, but it displays upvotes)

I tired to best to implement MVVM pattern in this app.

API used for this project: 
https://cat-fact.herokuapp.com

Bartozo committed:
Added Retrofit and serialization libraries

Bartozo committed:
Added UI libraries

-material
-recyclerview
-cardview

Bartozo committed:
Added coroutines library and missing apply plugin for kotlinx-serialization

Bartozo committed:
Added permissions required for Retrofit

- added internet permission
 - added network state permission

Bartozo committed:
Added data models and  basic networking classes.

-data models (CatFact, Name, User)
- networking classes (RemoteApiService, NetworkStatusChecker)

Bartozo committed:
Added the rest of networking classes

Bartozo committed:
Fixed issue with kotlin and serialization

-changed version to 1.3.61 from 1.3.72
-added  "nonstrict" to the converterFactory

Bartozo committed:
Added downloading data method to main acitvity

Bartozo committed:
Added layout in xml to the main activity

Bartozo committed:
Added RecyclerView with adapter and holder

Bartozo committed:
Fixed problem with downloading data when user was null

kotlinx.serialization.MissingFieldException: Field 'user' is required, but it was missing
- changed  type User to optional

Bartozo committed:
Added libraries for room database

Bartozo committed:
Added the rest od Database classes and ViewModel for CatFacts

-added lifecycle libraries
-added  daos and repository for db
- added view model

Bartozo committed:
Fixed the issue with room and view model 1/2

- couldnt create instance of view mode (fixedl
- couldnt add embedded value to room (fixed)

Bartozo committed:
Fixed the issue with room and view model 2/2

Bartozo committed:
Added glide library

Bartozo committed:
Added glide usage and changed the avatar imageView to be inside a cardView

Bartozo committed:
Added snackbars, diviider for recylcerview and removed unused code

filbabic commented on Week7/app/src/main/java/com/bartoszkrol/catfacts/App.kt:
```diff
@@ -0,0 +1,22 @@
+        private lateinit var instance: App
+
+        private val apiService by lazy { buildApiService() }
+
```
Would love to see the DAO and Repository here too

Bartozo replied:
I couldn't create a repository inside the application class. I think it's because database is no't ready at the start (problem with synchronization).

filbabic replied:
Well you can also use a lazy value, and have the context passed in be the `instance`.

It should work 100%, as that's usually the route I take!

Bartozo replied:
I have already tried using lazy just like in apiService. I get this error "lateinit property catdao has not been initialized". I am passing context from the instance class, i also tried to use lateinit var and init this classes inside onCreate method.

filbabic replied:
Well if you add it to the App class, and have everything as a lazy-delegate-value, including the DAO, you should be fine.

Bartozo replied:
val catdao by lazy {
            CatFactsDatabase.getDatabase(instance).catFactsDao()
        }
        
        val repository: RoomRepository by lazy {
            RoomRepository(catdao)
        }

Bartozo replied:
Caused by: kotlin.UninitializedPropertyAccessException: lateinit property instance has not been initialized
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:19)
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getCatdao(App.kt)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:23)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getRepository(App.kt)
        at com.bartoszkrol.catfacts.viewmodel.CatFactsViewModel.<init>(CatFactsViewModel.kt:15)

filbabic replied:
How do you use the repository in the ViewModel though?

Bartozo replied:
Inside the CatFactsViewModel:

private val repository: RoomRepository = App.repository
...
fun getCatFacts() = repository.getAllCatFacts()
fun removeCatFacts() = viewModelScope.launch {
        repository.removeAllCatFacts()
}
fun insertCatFacts(catFacts: List<CatFact>) = viewModelScope.launch {
        repository.insertCatFacts(catFacts)
}

filbabic replied:
You should either pass it in the constructor, or use a lazy delegate here too. Never use direct assigning, because that's memory hogging, and if you don't reach that screen/part of the app, it's just going to take up memory right away.

There might be another issue in the setup, but from the crash, it's coming from the ViewModel!

Bartozo replied:
I removed this field and changed it to repository directly from the application class.

class CatFactsViewModel(application: Application) : AndroidViewModel(application) {

    fun getCatFacts() = App.repository.getAllCatFacts()

    fun removeCatFacts() = viewModelScope.launch {
        App.repository.removeAllCatFacts()
    }

    fun insertCatFacts(catFacts: List<CatFact>) = viewModelScope.launch {
        App.repository.insertCatFacts(catFacts)
    }

}

Bartozo replied:
Still got an error

   Caused by: kotlin.UninitializedPropertyAccessException: lateinit property instance has not been initialized
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:19)
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getCatdao(App.kt)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:23)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getRepository(App.kt)
        at com.bartoszkrol.catfacts.viewmodel.CatFactsViewModel.getCatFacts(CatFactsViewModel.kt:13)
        at com.bartoszkrol.catfacts.ui.main.MainActivity.onCreate(MainActivity.kt:40)
        at android.app.Activity.performCreate(Activity.java:6237)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)

filbabic replied:
That's also not good, because you won't be able to test this.

I did mention the best way to go about this is to use the constructor to send the repository, and not having to create it here. Try if you can do it that way!

Bartozo replied:
How can i pass the constructor to the viewmodel when i am initializing it like this 
    private val catFactsViewModel: CatFactsViewModel by lazy {
        ViewModelProvider(this).get(CatFactsViewModel::class.java)
    }

filbabic replied:
https://medium.com/koderlabs/viewmodel-with-viewmodelprovider-factory-the-creator-of-viewmodel-8fabfec1aa4f

We linked this article in the assignments, you should be able to use the factory to pass in everything.

Let me know if this works!

Bartozo replied:
I have created this factory for my viewmodel but the error still remains the same. Right now i am debugging to get more information why this isn't working.

filbabic replied:
I will try to check out the project a bit later and check it out for you too, if you don't find out the issue cause

Bartozo replied:
Yey! I found the solution :)

Bartozo replied:
Debugging the app really helped me a lot! I found that my properties inside the application class weren't ready before my activity was created.  I checked manifest file and i was missing my application class tag

android:name=".App"

filbabic replied:
Awesome! Glad we worked it out! :]

Bartozo replied:
Thanks for your help filip!
I have fixed the code so I can merge PR

filbabic commented on Week7/app/src/main/java/com/bartoszkrol/catfacts/viewmodel/CatFactsViewModel.kt:
```diff
@@ -0,0 +1,35 @@
+    private val allCatFacts: LiveData<List<CatFact>>
+
+    init {
+        val catFactsDao = CatFactsDatabase.getDatabase(application).catFactsDao()
```
Avoid initializing dependencies in classes that use them. If they change, you have to change all the classes too.

Use the App class to provide dependencies instead, and pass them in through the constructor!

filbabic reviewed (approved):
Not much to say, nice PR with only a few changes, and very well structured code!

I left some comments regarding the initialization logic for your ViewModels, but that's about it!

Nice job, everything's really clear and well structured! I'll give you **Above & Beyond** cause of the clarity and structure!

Bartozo committed:
Fixed some things in my assignment

- added missing app in manifest
- added viewModelFactory for CatFactsViewModel
- moved repository and dao to be in the application object
